srml/generic-asset/src/lib.rs

```diff
@@ -0,0 +1,1241 @@
+				.checked_add(&One::one())
+				.ok_or_else(|| "No new user asset id available.")?;
+			<NextAssetId<T>>::put(next_id);
+			asset_id
+		};
+
+		let account_id = from_account.unwrap_or_default();
+		let permissions: PermissionVersions<T::AccountId> = options.permissions.clone().into();
+
+		<TotalIssuance<T>>::insert(asset_id, &options.initial_issuance);
+		<FreeBalance<T>>::insert(&asset_id, &account_id, options.initial_issuance);
+		<Permissions<T>>::insert(&asset_id, permissions);
+
+		Self::deposit_event(RawEvent::Created(asset_id, account_id, options));
+
+		Ok(())
+	}
+
+	/// Transfer some liquid free balance from one account to another.
+	/// This will not emit the `Transferred` event.
+	pub fn make_transfer(asset_id: &T::AssetId, from: &T::AccountId, to: &T::AccountId, amount: T::Balance) -> Result {
+		let new_balance = Self::free_balance(asset_id, from)
+			.checked_sub(&amount)
+			.ok_or_else(|| "balance too low to send amount")?;
+		Self::ensure_can_withdraw(asset_id, from, amount, WithdrawReason::Transfer, new_balance)?;
+
+		if from != to {
+			<FreeBalance<T>>::mutate(asset_id, from, |balance| *balance -= amount);
+			<FreeBalance<T>>::mutate(asset_id, to, |balance| *balance += amount);
+		}
+
+		Ok(())
+	}
+
+	/// Transfer some liquid free balance from one account to another.
+	/// This will emit the `Transferred` event.
+	pub fn make_transfer_with_event(
+		asset_id: &T::AssetId,
+		from: &T::AccountId,
+		to: &T::AccountId,
+		amount: T::Balance,
+	) -> Result {
+		Self::make_transfer(asset_id, from, to, amount)?;
+
+		if from != to {
+			Self::deposit_event(RawEvent::Transferred(*asset_id, from.clone(), to.clone(), amount));
+		}
+
+		Ok(())
+	}
+
+	/// Move `amount` from free balance to reserved balance.
+	///
+	/// If the free balance is lower than `amount`, then no funds will be moved and an `Err` will
+	/// be returned. This is different behavior than `unreserve`.
+	pub fn reserve(asset_id: &T::AssetId, who: &T::AccountId, amount: T::Balance) -> Result {
+		// Do we need to consider that this is an atomic transaction?
+		let original_reserve_balance = Self::reserved_balance(asset_id, who);
+		let original_free_balance = Self::free_balance(asset_id, who);
+		if original_free_balance < amount {
+			return Err("not enough free funds");
+		}
+		let new_reserve_balance = original_reserve_balance + amount;
+		Self::set_reserved_balance(asset_id, who, new_reserve_balance);
+		let new_free_balance = original_free_balance - amount;
+		Self::set_free_balance(asset_id, who, new_free_balance);
+		Ok(())
+	}
+
+	/// Moves up to `amount` from reserved balance to free balance. This function cannot fail.
+	///
+	/// As many assets up to `amount` will be moved as possible. If the reserve balance of `who`
+	/// is less than `amount`, then the remaining amount will be returned.
+	/// NOTE: This is different behavior than `reserve`.
+	pub fn unreserve(asset_id: &T::AssetId, who: &T::AccountId, amount: T::Balance) -> T::Balance {
```

Comment:
this should not be public, since it doesn't handle imbalances.

Reply:
doesn't modify imbalances/total issuances, it just move fund from reserved balances to free balances.